File: sdks/python/src/fyredocs/_billing.py
"""``client.billing`` — billing-service /api/billing/v1/billing/* wrapper."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ._types import (
    BillingMeResponse,
    Plan,
    SubscribeRequest,
    Subscription,
    _billing_me_from_json,
    _plan_from_json,
    _subscription_from_json,
)

if TYPE_CHECKING:
    from ._client import Client


class BillingAPI:
    """Wraps ``/api/billing/v1/billing/*``."""

    def __init__(self, client: "Client") -> None:
        self._c = client

    def plans(self) -> list[Plan]:
        """List self-serve + sales-led plans the user is eligible
        to subscribe to."""
        data = self._c.request("/api/billing/v1/billing/plans")
        if not isinstance(data, dict):
            return []
        return [_plan_from_json(p) for p in (data.get("plans") or [])]

    def me(self) -> BillingMeResponse:
        """Return the calling user's plan + subscription + current
        usage rollup."""
        data = self._c.request("/api/billing/v1/billing/me")
        if not isinstance(data, dict):
            raise ValueError("billing.me: unexpected response shape")
        return _billing_me_from_json(data)

    def subscribe(self, req: SubscribeRequest) -> Subscription:
        """Upgrade / downgrade / start a subscription. Returns the
        updated Subscription."""
        body: dict[str, object] = {"planCode": req.plan_code}
        if req.seats is not None:
            body["seats"] = req.seats
        data = self._c.request(
            "/api/billing/v1/billing/me/subscribe", method="POST", body=body,
        )
        if not isinstance(data, dict):
            raise ValueError("billing.subscribe: unexpected response shape")
        return _subscription_from_json(data)

File: sdks/go/billing.go
package fyredocs

import (
	"context"
	"net/http"
)

// BillingAPI wraps the `/api/billing/v1/billing/*` endpoints owned
// by billing-service.
type BillingAPI struct {
	c *Client
}

// Plans returns the list of self-serve + sales-led plans the user
// is eligible to subscribe to.
func (b *BillingAPI) Plans(ctx context.Context) ([]Plan, error) {
	var wrapper struct {
		Plans []Plan `json:"plans"`
	}
	if err := b.c.Request(ctx, "/api/billing/v1/billing/plans", RequestOptions{
		Method: http.MethodGet,
		Out:    &wrapper,
	}); err != nil {
		return nil, err
	}
	return wrapper.Plans, nil
}

// Me returns the calling user's plan + subscription + current
// usage rollup.
func (b *BillingAPI) Me(ctx context.Context) (*BillingMeResponse, error) {
	var out BillingMeResponse
	if err := b.c.Request(ctx, "/api/billing/v1/billing/me", RequestOptions{
		Method: http.MethodGet,
		Out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

// Subscribe upgrades / downgrades / starts the calling user's
// subscription. Returns the updated Subscription.
func (b *BillingAPI) Subscribe(ctx context.Context, req SubscribeRequest) (*Subscription, error) {
	var out Subscription
	if err := b.c.Request(ctx, "/api/billing/v1/billing/me/subscribe", RequestOptions{
		Method: http.MethodPost,
		Body:   req,
		Out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}
